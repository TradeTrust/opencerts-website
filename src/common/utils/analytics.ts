export const validateGtag = function () {
  const isInit = typeof gtag === "function";
  if (!isInit) console.warn("gtag is not initialised");
  return isInit;
};
export interface GAEvent {
  action: string;
  category?: string;
  label?: string;
  value?: number;
}

export const validatePageViewEvent = function (gaEvent: GAEvent) {
  const action = gaEvent.action;
  if (!action) console.error("Action is required");
  if (action && typeof action !== "string") console.error("Action must be a string");
};
export const gaPageView = function (gaEvent: GAEvent, gaId: string) {
  if (!validateGtag()) return;
  validatePageViewEvent(gaEvent);
  const action: string = gaEvent.action;
  gtag("event", action, {
    send_to: gaId,
  });
};
export const validateGaEvent = function (gaEvent: GAEvent) {
  const action = gaEvent.action,
    category = gaEvent.category,
    label = gaEvent.label,
    value = gaEvent.value;
  if (!category) console.error("Category is required");
  if (!action) console.error("Action is required");
  if (label && typeof label !== "string") console.error("Label must be a string");
  if (value && typeof value !== "number") console.error("Value must be a number");
};
export const gaEvent = function (eventData: GAEvent) {
  if (!validateGtag()) return;
  validateGaEvent(eventData);
  const action = eventData.action,
    category = eventData.category,
    label = eventData.label,
    value = eventData.value;
  gtag("event", action, {
    event_category: category,
    event_label: label,
    value: value,
  });
};
