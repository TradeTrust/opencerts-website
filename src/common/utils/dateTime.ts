import { format, utcToZonedTime } from "date-fns-tz";

export const formatTime = (time: string, dateFormat: string, timeZoneSg = "Asia/Singapore"): string => {
  const dateTimeSg = utcToZonedTime(time, timeZoneSg);
  return format(dateTimeSg, dateFormat, { timeZone: timeZoneSg });
};
