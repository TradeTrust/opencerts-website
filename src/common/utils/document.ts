import { WrappedDocument, getData, v2 } from "@govtechsg/open-attestation";

export const getDocumentId = (document: WrappedDocument) => `0x${document.signature.targetHash}`;
export const getTokenRegistryAddress = (document: WrappedDocument<v2.OpenAttestationDocument>) =>
  getData(document).issuers[0].tokenRegistry;
