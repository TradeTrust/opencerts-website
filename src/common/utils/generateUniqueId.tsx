export const generateUniqueId = () => {
  return "_" + Math.random().toString(36).substr(2, 9);
}; // https://gist.github.com/gordonbrander/2230317
