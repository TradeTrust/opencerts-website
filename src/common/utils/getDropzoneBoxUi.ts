enum DropzoneBoxUiState {
  DRAG_REJECT = "border-scarlet-400 bg-scarlet-100",
  DRAG_ACTIVE = "border-forest-200 bg-emerald-50",
  DRAG_ACCEPT = "border-forest-200 bg-emerald-50",
  VERIFICATION_PENDING = "border-cloud-100 bg-white",
  VERIFICATION_ERROR = "border-scarlet-400 bg-scarlet-100",
  ACTION_ERROR = "border-scarlet-400 bg-scarlet-100",
  DEFAULT = "border-cloud-100 bg-white",
}

interface GetDropzoneBoxUi {
  isDragReject: boolean;
  isDragActive: boolean;
  isDragAccept: boolean;
  isVerificationPending: boolean;
  isVerificationError: boolean | null;
  isActionError?: boolean | null;
}

export const getDropzoneBoxUi = ({
  isDragReject,
  isDragActive,
  isDragAccept,
  isVerificationPending,
  isVerificationError,
  isActionError,
}: GetDropzoneBoxUi): string => {
  switch (true) {
    case isDragReject:
      return DropzoneBoxUiState.DRAG_REJECT;
    case isDragActive:
      return DropzoneBoxUiState.DRAG_ACTIVE;
    case isDragAccept:
      return DropzoneBoxUiState.DRAG_ACCEPT;
    case isVerificationPending:
      return DropzoneBoxUiState.VERIFICATION_PENDING;
    case isVerificationError:
      return DropzoneBoxUiState.VERIFICATION_ERROR;
    case isActionError:
      return DropzoneBoxUiState.ACTION_ERROR;
    default:
      return DropzoneBoxUiState.DEFAULT;
  }
};
