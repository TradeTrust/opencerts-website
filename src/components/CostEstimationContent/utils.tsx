import {
  TRANSFER_OWNERSHIP_GAS,
  TRANSFER_HOLDERSHIP_GAS,
  ISSUE_DOC_GAS,
  BURN_DOC_GAS,
  SURRENDER_DOC_GAS,
} from "../../constants/cost-estimation";
import { CostData } from "./types";
import { PersonaContractCost, PersonaCostInformation, Persona } from "../../models/CostPersona";

const getGasPrice = (gas: number, priceFactor: number): number => gas * priceFactor;

export const makeExporter = (priceFactor: number, maticPriceFactor: number, description: string): CostData => {
  const transferHoldershipContractCost = new PersonaContractCost(
    "/static/images/cost/exporter/transfer-holdership-icon.png",
    "Cost to transfer holdership",
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, maticPriceFactor)}`
  );
  const transferOwnershipContractCost = new PersonaContractCost(
    "/static/images/cost/exporter/transfer-ownership-icon.png",
    "Cost to transfer ownership",
    `${getGasPrice(TRANSFER_OWNERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_OWNERSHIP_GAS, maticPriceFactor)}`
  );
  const totalContractCost = new PersonaContractCost(
    "/static/images/cost/total-cost-icon.png",
    "Total Cost",
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, priceFactor) + getGasPrice(TRANSFER_OWNERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, maticPriceFactor) + getGasPrice(TRANSFER_OWNERSHIP_GAS, maticPriceFactor)}`
  );
  const exporterCostInformation = new PersonaCostInformation(
    "Cost for The Exporter",
    [transferHoldershipContractCost, transferOwnershipContractCost, totalContractCost],
    description
  );
  return new Persona("The Exporter", "/static/images/home/howItWorks/persona/persona1.png", exporterCostInformation);
};

export const makeCarrier = (priceFactor: number, maticPriceFactor: number, description: string): CostData => {
  const issueContractCost = new PersonaContractCost(
    "/static/images/cost/carrier/issue-ebl-icon.png",
    "Cost to Issue eBL",
    `${getGasPrice(ISSUE_DOC_GAS, priceFactor)}`,
    `${getGasPrice(ISSUE_DOC_GAS, maticPriceFactor)}`
  );
  const burnContractCost = new PersonaContractCost(
    "/static/images/cost/carrier/burn-ebl-icon.png",
    "Cost to Burn eBL",
    `${getGasPrice(BURN_DOC_GAS, priceFactor)}`,
    `${getGasPrice(BURN_DOC_GAS, maticPriceFactor)}`
  );
  const totalContractCost = new PersonaContractCost(
    "/static/images/cost/total-cost-icon.png",
    "Total Cost",
    `${getGasPrice(ISSUE_DOC_GAS, priceFactor) + getGasPrice(BURN_DOC_GAS, priceFactor)}`,
    `${getGasPrice(ISSUE_DOC_GAS, maticPriceFactor) + getGasPrice(BURN_DOC_GAS, maticPriceFactor)}`
  );
  const carrierCostInformation = new PersonaCostInformation(
    "Cost for The Exporter",
    [issueContractCost, burnContractCost, totalContractCost],
    description
  );
  return new Persona("The Carrier", "/static/images/home/howItWorks/persona/persona2.png", carrierCostInformation);
};

export const makeImporter = (priceFactor: number, maticPriceFactor: number, description: string): CostData => {
  const surrenderContractCost = new PersonaContractCost(
    "/static/images/cost/importer/surrender-ebl-icon.png",
    "Surrender eBL",
    `${getGasPrice(SURRENDER_DOC_GAS, priceFactor)}`,
    `${getGasPrice(SURRENDER_DOC_GAS, maticPriceFactor)}`
  );
  const importerCostInformation = new PersonaCostInformation(
    "Cost for The Exporter",
    [surrenderContractCost],
    description
  );
  return new Persona("The Importer", "/static/images/home/howItWorks/persona/persona3.png", importerCostInformation);
};

export const makeBanker = (priceFactor: number, maticPriceFactor: number, description: string): CostData => {
  const transferHoldershipContractCost = new PersonaContractCost(
    "/static/images/cost/exporter/transfer-holdership-icon.png",
    "Cost to transfer holdership",
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, maticPriceFactor)}`
  );
  const transferOwnershipContractCost = new PersonaContractCost(
    "/static/images/cost/exporter/transfer-ownership-icon.png",
    "Cost to transfer ownership",
    `${getGasPrice(TRANSFER_OWNERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_OWNERSHIP_GAS, maticPriceFactor)}`
  );
  const totalContractCost = new PersonaContractCost(
    "/static/images/cost/total-cost-icon.png",
    "Total Cost",
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, priceFactor) + getGasPrice(TRANSFER_OWNERSHIP_GAS, priceFactor)}`,
    `${getGasPrice(TRANSFER_HOLDERSHIP_GAS, maticPriceFactor) + getGasPrice(TRANSFER_OWNERSHIP_GAS, maticPriceFactor)}`
  );
  const bankerCostInformation = new PersonaCostInformation(
    "Cost for The Exporter",
    [transferHoldershipContractCost, transferOwnershipContractCost, totalContractCost],
    description
  );
  return new Persona("The Banker", "/static/images/home/howItWorks/persona/persona4.png", bankerCostInformation);
};
