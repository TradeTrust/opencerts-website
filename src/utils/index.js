import { hexToNumberString } from "web3-utils";
import { getData } from "@govtechsg/open-attestation";
import { ETHERSCAN_BASE_URL } from "../config";

const ethereumAddressMatcher = /^0x[a-fA-F0-9]{40}$/;
export function isEthereumAddress(address) {
  return ethereumAddressMatcher.test(address);
}

export const makeEtherscanAddressURL = (address) => {
  return `${ETHERSCAN_BASE_URL}address/${address}`;
};

export const makeEtherscanTokenURL = ({ registryAddress, tokenId }) => {
  const tokenIdDecimal = hexToNumberString(`0x${tokenId}`);
  return `${ETHERSCAN_BASE_URL}token/${registryAddress}?a=${tokenIdDecimal}`;
};

export const getAssetInfo = (document) => {
  const { tokenRegistry } = getData(document).issuers[0];
  const { merkleRoot: tokenId } = document.signature;
  return { tokenRegistry, tokenId };
};
