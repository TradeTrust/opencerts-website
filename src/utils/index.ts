import { utils } from "ethers";
import { ETHERSCAN_BASE_URL } from "../config";
import { compareDesc, compareAsc } from "date-fns";

export const makeEtherscanAddressURL = (address: string): string => {
  return `${ETHERSCAN_BASE_URL}address/${address}`;
};

export const isEthereumAddress = (address: string): boolean | undefined => {
  try {
    if (utils.getAddress(address)) {
      return true;
    }
  } catch (e: any) {
    if (e.reason === "invalid address") {
      return false;
    } else throw e;
  }
};

export const convertSecondsToMinAndSec = (seconds: number): string => {
  const sec = seconds % 60;
  return `${~~(seconds / 60)}:${sec < 10 ? `0${sec}` : sec}m`;
};

export const getSortedByDateDesc = (items: any[]): any[] => {
  items.sort((a, b): number => {
    return compareDesc(new Date(a.attributes.date), new Date(b.attributes.date));
  });

  return items;
};

export const getSortedByDateAsc = (items: any[]): any[] => {
  items.sort((a, b): number => {
    return compareAsc(new Date(a.attributes.date), new Date(b.attributes.date));
  });

  return items;
};

// https://docs.netlify.com/forms/setup/#submit-javascript-rendered-forms-with-ajax
export const encode: any = (data: { [x: string]: string | number | boolean }) => {
  return Object.keys(data)
    .map((key) => encodeURIComponent(key) + "=" + encodeURIComponent(data[key]))
    .join("&");
};

export const addClassNameIfExist = (className?: string): string => {
  if (!className) {
    return "";
  }

  return className;
};
