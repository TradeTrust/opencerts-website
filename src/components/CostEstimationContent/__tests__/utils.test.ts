import { getGasPrice, makeExporter, makeBanker, makeCarrier, makeImporter } from "../utils";

describe("getGasPrice", () => {
  it("should return the 'gas price' correctly", () => {
    expect(getGasPrice(463, 18)).toEqual(8334);
  });
});

const description = "This is a mock description";

describe("makeExporter", () => {
  it("should make The Exporter persona data correctly", () => {
    const exporter = makeExporter(8334, 180, description);
    expect(exporter).toMatchSnapshot();
  });
});
describe("makeImporter", () => {
  it("should make The Importer persona data correctly", () => {
    const importer = makeImporter(8334, 180, description);
    expect(importer).toMatchSnapshot();
  });
});
describe("makeBanker", () => {
  it("should make The Banker persona data correctly", () => {
    const banker = makeBanker(8334, 180, description);
    expect(banker).toMatchSnapshot();
  });
});
describe("makeCarrier", () => {
  it("should make The Carrier persona data correctly", () => {
    const carrier = makeCarrier(8334, 180, description);
    expect(carrier).toMatchSnapshot();
  });
});
