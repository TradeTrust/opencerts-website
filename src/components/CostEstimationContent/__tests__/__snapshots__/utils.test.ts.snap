// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeBanker should make The Banker persona data correctly 1`] = `
Persona {
  "costInformation": PersonaCostInformation {
    "contractCost": Array [
      PersonaContractCost {
        "contractIcon": "/static/images/cost/exporter/transfer-holdership-icon.png",
        "contractPriceInEthereum": "363645756",
        "contractPriceInPolygon": "7854120",
        "contractTitle": "Cost to transfer holdership",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/exporter/transfer-ownership-icon.png",
        "contractPriceInEthereum": "2705949792",
        "contractPriceInPolygon": "58443840",
        "contractTitle": "Cost to transfer ownership",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/total-cost-icon.png",
        "contractPriceInEthereum": "3069595548",
        "contractPriceInPolygon": "66297960",
        "contractTitle": "Total Cost",
      },
    ],
    "costTitle": "Cost for The Exporter",
    "description": "This is a mock description",
  },
  "icon": "/static/images/home/howItWorks/persona/persona4.png",
  "jobTitle": "The Banker",
}
`;

exports[`makeCarrier should make The Carrier persona data correctly 1`] = `
Persona {
  "costInformation": PersonaCostInformation {
    "contractCost": Array [
      PersonaContractCost {
        "contractIcon": "/static/images/cost/carrier/issue-ebl-icon.png",
        "contractPriceInEthereum": "1996184682",
        "contractPriceInPolygon": "43114140",
        "contractTitle": "Cost to Issue eBL",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/carrier/burn-ebl-icon.png",
        "contractPriceInEthereum": "471137688",
        "contractPriceInPolygon": "10175760",
        "contractTitle": "Cost to Burn eBL",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/total-cost-icon.png",
        "contractPriceInEthereum": "2467322370",
        "contractPriceInPolygon": "53289900",
        "contractTitle": "Total Cost",
      },
    ],
    "costTitle": "Cost for The Exporter",
    "description": "This is a mock description",
  },
  "icon": "/static/images/home/howItWorks/persona/persona2.png",
  "jobTitle": "The Carrier",
}
`;

exports[`makeExporter should make The Exporter persona data correctly 1`] = `
Persona {
  "costInformation": PersonaCostInformation {
    "contractCost": Array [
      PersonaContractCost {
        "contractIcon": "/static/images/cost/exporter/transfer-holdership-icon.png",
        "contractPriceInEthereum": "363645756",
        "contractPriceInPolygon": "7854120",
        "contractTitle": "Cost to transfer holdership",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/exporter/transfer-ownership-icon.png",
        "contractPriceInEthereum": "2705949792",
        "contractPriceInPolygon": "58443840",
        "contractTitle": "Cost to transfer ownership",
      },
      PersonaContractCost {
        "contractIcon": "/static/images/cost/total-cost-icon.png",
        "contractPriceInEthereum": "3069595548",
        "contractPriceInPolygon": "66297960",
        "contractTitle": "Total Cost",
      },
    ],
    "costTitle": "Cost for The Exporter",
    "description": "This is a mock description",
  },
  "icon": "/static/images/home/howItWorks/persona/persona1.png",
  "jobTitle": "The Exporter",
}
`;

exports[`makeImporter should make The Importer persona data correctly 1`] = `
Persona {
  "costInformation": PersonaCostInformation {
    "contractCost": Array [
      PersonaContractCost {
        "contractIcon": "/static/images/cost/importer/surrender-ebl-icon.png",
        "contractPriceInEthereum": "778687290",
        "contractPriceInPolygon": "16818300",
        "contractTitle": "Surrender eBL",
      },
    ],
    "costTitle": "Cost for The Exporter",
    "description": "This is a mock description",
  },
  "icon": "/static/images/home/howItWorks/persona/persona3.png",
  "jobTitle": "The Importer",
}
`;
